saleor/checkout/utils.py

```diff
@@ -81,6 +114,61 @@ def invalidate_checkout_prices(
     return updated_fields
 
 
+def apply_gift_reward_if_applicable(checkout: "Checkout") -> None:
+    """Apply gift reward if applicable on newly created checkout.
+
+    This method apply the gift reward if any gift promotion exists and
+    when it's giving the best discount on the current checkout.
+    """
+    if not PromotionRule.objects.filter(gifts__isnull=False).exists():
+        return
+
+    variants_id = [line.variant_id for line in checkout.lines.all()]
+
+    # calculate base subtotal and total
+    variant_id_to_discounted_price = {
+        variant_id: discounted_price or price
+        for variant_id, discounted_price, price in product_models.ProductVariantChannelListing.objects.filter(
+            variant_id__in=variants_id,
+            channel_id=checkout.channel_id,
+        ).values_list("variant_id", "discounted_price_amount", "price_amount")
+    }
+    subtotal = Decimal("0")
+    for line in checkout.lines.all():
+        if price_amount := line.price_override:
+            price = price_amount
+        else:
+            price = variant_id_to_discounted_price.get(line.variant_id) or Decimal("0")
+        subtotal += price * line.quantity
+    checkout.base_subtotal = Money(subtotal, checkout.currency)
+    # base total and subtotal is the same, as there is no option to set the
+    # delivery method during checkout creation
+    checkout.base_total = checkout.base_subtotal
+    checkout.save(update_fields=["base_subtotal_amount", "base_total_amount"])
```

Comment:
Maybe we could wrapt it to some internal function?

Reply:
Done